coreax/solvers/coresubset.py

```diff
@@ -458,3 +486,301 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
+            # zeros simultaneously.
+            feature_gramian = jnp.pad(feature_gramian, [(0, 1)], mode="constant")
+            response_gramian = jnp.pad(response_gramian, [(0, 1)], mode="constant")
+            training_cme = jnp.pad(training_cme, [(0, 1)], mode="constant")
+
+        else:
+            feature_gramian = solver_state.feature_gramian
+            response_gramian = solver_state.response_gramian
+            training_cme = solver_state.training_cme
+
+        # Sample the indices to be considered at each iteration ahead of time.
+        if self.batch_size is not None and self.batch_size < num_data_pairs:
+            batch_size = self.batch_size
+        else:
+            batch_size = num_data_pairs
+        batch_indices = sample_batch_indices(
+            random_key=self.random_key,
+            max_index=num_data_pairs,
+            batch_size=batch_size,
+            num_batches=self.coreset_size,
+        )
+
+        # Insert coreset indices into the batch indices to avoid degrading the CMMD
+        # (only has an effect when refining)
+        if self.batch_size is not None:
+            batch_indices = batch_indices.at[:, -1].set(coreset_indices)
+
+        # Initialise an array that will let us extract and build rectangular arrays
+        # which stack arrays corresponding to every possible candidate coreset.
+        initial_candidate_coresets = (
+            jnp.tile(coreset_indices, (batch_size, 1)).at[:, 0].set(batch_indices[0, :])
+        )
+
+        # Adaptively initialise an "identity matrix". For reduction we need this to
+        # be a zeros matrix which we will iteratively add ones to on the diagonal. While
+        # for refinement we need an actual identity matrix.
+        identity_helper = jnp.hstack((jnp.ones(num_data_pairs), jnp.array([0])))
+        identity = jnp.diag(identity_helper[coreset_indices])
+
+        def _greedy_body(
+            i: int, val: tuple[Array, Array, Array]
+        ) -> tuple[Array, Array, Array]:
+            r"""Execute main loop of GreedyCMMD."""
+            coreset_indices, identity, candidate_coresets = val
+
+            # Update the identity matrix to allow for sub-array inversion in the
+            # case of reduction (no effect when refining)
+            updated_identity = identity.at[i, i].set(1)
+
+            # Compute the loss (CMMD) corresponding to each candidate coreset
+            loss = _greedy_cmmd_loss(
+                self.random_key,
+                candidate_coresets,
+                feature_gramian,
+                response_gramian,
+                training_cme,
+                self.regularisation_parameter,
+                updated_identity,
+                self.inverse_approximator,
+            )
+
+            # Find the optimal replacement coreset index, ensuring we don't pick an
+            # already chosen point if we want the indices to be unique.
+            if self.unique:
```

Comment:
I think all unit tests have unique as True - if I am correct it might be worth testing this case where unique is false